### PR TITLE
Parse definers

### PR DESCRIPTION
**Pull request by muhamadSammar**

### CODE REVIEW

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 1]
```diff
@@ -281,13 +334,23 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
...
+
+def mlx5_ste_v1_tag_parser(entry_format, definer_id, tag, raw):
+        func, inner = None, None
 
-	if entry_format not in switch_tag_parser.keys():
-		print("Err: Unsupported STEv1 tag format")
-		return {}
+	if entry_format in switch_tag_parser.keys():
+	    func, inner = switch_tag_parser[entry_format]
+        elif definer_id not in [None, -1, "-1"] and int(definer_id, 16) in switch_definer_parser.keys():
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:347]: add comment explaning differences between None, -1, "-1"

Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:347]: weird indentation

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 2]
```diff
@@ -281,13 +334,23 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
...
+            definer_id = int(definer_id, 16)
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:348]: don't override the func param, just do this on the next line

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 3]
```diff
@@ -281,13 +334,23 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
...
+            func = switch_definer_parser[definer_id]
+        else:
+            print("Err: Unsupported STEv1 tag format, entry_format=%s, definer_id=%s" %(entry_format,definer_id))
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:351]: add space after comma

**Review by dariavel — CHANGES_REQUESTED** [review]

File: src/dr_action.py [hunk 4]
```diff
@@ -40,7 +40,7 @@ def dr_rec_type_is_action(rec_type):
 class dr_dump_action_drop(dr_obj):
     def __init__(self, data):
         keys = ["dr_dump_rec_type", "id", "rule_id"]
-        self.data = dict(zip(keys, data))
+        self.data = dict(zip(keys, data + [None]*(len(keys)-len(data))))
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply src/dr_action.py:43]: done

File: src/dr_action.py [hunk 5]
```diff
@@ -96,10 +96,10 @@ def dump_str(self):
 class dr_dump_action_modify_header(dr_obj):
     def __init__(self, data):
         keys = ["dr_dump_rec_type", "id", "rule_id", "rewrite_index", "single_action_opt"]
-        self.data = dict(zip(keys, data))
+        self.data = dict(zip(keys, data + [None]*(len(keys)-len(data))))
 
     def dump_str(self):
-        if "single_action_opt" in self.data.keys():
+        if "single_action_opt" in self.data.keys() and self.data["single_action_opt"] is not None:
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply src/dr_action.py:102]: done

File: src/parsers/dr_ste_v1_parser.py [hunk 6]
```diff
@@ -103,26 +103,29 @@ def mlx5_ifc_ste_v1_match_bits_parser(bin_str, raw) :
...
-
-    ret["tag"] = {"info" : "STE only contains actions"}
-
+  
+    if len(bin_str) == 512:
+        tag = bin_str[256 : 512]
+        lookup_type = int(bin_str[0 : 8] + bin_str[48 : 56], 2)
+        ret["tag"] = mlx5_ste_v1_tag_parser(lookup_type, definer_id, tag, raw)
+    else:
+        ret["tag"] = {"tag": "failed to parse"}
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_parser.py:112]: done

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 7]
```diff
@@ -281,13 +334,23 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
     }
 
 
-def mlx5_ste_v1_tag_parser(entry_format, tag, raw):
+switch_definer_parser = {
+    0x22: mlx5_ifc_ste_def22_v1_bits_tag_parser,
+    0x24: mlx5_ifc_ste_def24_v1_bits_tag_parser, 
+}
+
+def mlx5_ste_v1_tag_parser(entry_format, definer_id, tag, raw):
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:342]: done

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 1 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:347]: done

Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:347]: done

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 2 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py:348]: done

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 8]
```diff
@@ -281,19 +334,30 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
...
+    0x24: mlx5_ifc_ste_def24_v1_bits_tag_parser, 
+}
+
+def mlx5_ste_v1_tag_parser(lookup_type, definer_id, tag, raw):
+        func, inner = None, None
+
+        if lookup_type in switch_tag_parser.keys():
+            func, inner = switch_tag_parser[lookup_type]
+        # The default value for definer_id is None, and value -1 means definer not supported.
+        elif definer_id not in [None,"-1"] and int(definer_id, 16) in switch_definer_parser.keys():
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:348]: `[None, "-1"]`
add space after comma

**Review by dariavel — CHANGES_REQUESTED** [review]

File: src/parsers/dr_ste_v1_tag_parser.py (hunk 8 above)
Comment (muhamadSammar) [reply src/parsers/dr_ste_v1_tag_parser.py]: done

File: src/parsers/dr_ste_v1_tag_parser.py [hunk 9]
```diff
@@ -281,19 +334,30 @@ def mlx5_ifc_ste_src_gvmi_qp_v1_bits_tag_parser(bin_str):
     }
 
 
-def mlx5_ste_v1_tag_parser(entry_format, tag, raw):
+switch_definer_parser = {
+    0x22: mlx5_ifc_ste_def22_v1_bits_tag_parser,
+    0x24: mlx5_ifc_ste_def24_v1_bits_tag_parser,
```
Comment (dariavel) [inline src/parsers/dr_ste_v1_tag_parser.py:339]: there is an extra space at the end of 0x24

**Review by dariavel — CHANGES_REQUESTED** [review]

**Review by dariavel — APPROVED** [review]

**Pull request merged**